### DISCUSSION TITLE
Color molecule with arbitrary array using API

### DISCUSSION THREAD

**Discussion by nilshof01** [BradyAJohnston/MolecularNodes#546] [Q&A]
Hi, 

I am trying to customize my molecules a little bit and would like to integrate that into a python pipeline. As a starting point I would like to change the color of the molecule where different parts have varying colors. I wanted to ask for the feasibility of that, since I tried to find a solution but was not very successful with that.
1. My first attempy included to just fetch the molecule and the use the b_factor column and assign it to the Color attribute of the corresponding molecule. However, the corresponding arrays have different shapes as shown here: 
```python
>>> import molecularnodes as mn
>>> mol1 = mn.io.fetch("7eow")
>>> mol1.get_attribute("Color").shape[0]
2600
>>> mol1.array.b_factor.shape[0]
2980

```
Thus, I get an error when I try to assign the b_factor column to the Color attribute with set_attributes. 

In another attempt I tried to just use the color flag from the fetch function but the color does not really change even if I use the create_starting_node_tree function directly with the following code snippets:

```python
>>> mol1 = mn.io.fetch("7eow", format = "pdb", cache_dir = r"C:\Users\nilsh\OneDrive\Desktop\ar_app", color = "plddt")
```

```python
>>> mn.blender.nodes.create_starting_node_tree(mol1.object, style = "cartoon", color = "plddt")
```

Nothing really happens for both codes. So the general goal is to parse an array with values and the molecule is colored differently according to this array.  Do you have a hint what I could do? :)

Thanks!

**Comment by BradyAJohnston** [comment]
Hello! 

I have to at first apologise for the state of the python API & the documentation around it. The plan is to _eventually_ have everything be scriptable and easier to interact with - but currently it's all a bit messy.

In short - yes this is very possible. How you would go about this is changing, with 4.2 coming for Blender 4.2 in the next week (#501).

There are two different sets of data that we are working with inside of Blender / Molecular Nodes.  There is the molecular data, that is accessible currently through the `mol.array`, which is a `biotite.structure.AtomArray` that resulted from parsing the file itself. 

Once an `Object` is created in the 3D scene inside of Blender, then we can access the attributes of the object itself with the `mol.get_attribute()` method. This is _confusingly named_ (currently). 

The reason you are seeing different lengths of the arrays, is that there has been some filtering applied when creating the object (removing solvent / water atoms).

```python
>>> m = mn.io.fetch('7eow')
>>> m.get_attribute('b_factor').shape
(2600,)

>>> m.get_attribute('Color').shape
(2600, 4)

>>> m.array.b_factor.shape
(2980,)
```

This has been discussed a bit in the #359 and like I said is being majorly overhauled in the 4.2 update #501. There is still lots to be improved, and I especially welcome and want to hear from people on use cases, and how they would like to interact with the data.

What is going to change for 4.2 is that accessing attributes on the object will become the following for reading and writing from the 3D object. This brings it to better align with the node names inside of Blender / Geometry Nodes and better distinguishes it from python attributes.

```py
mol.named_attribute()
mol.store_named_attribute()
```

The idea will be that you can generate some color values using python and store them on the Object which will be used when creating the 3D mesh.

```py
mol = mn.io.fetch('7eow')
mol.create_object(color=None) # this would stop the overriding of colors in the node tree

new_colors = some_color_function(mol.named_attribute('b_factor'))
mol.store_named_attribute(new_colors, 'Color')
```

Currently the `color` argument in the `mn.io.fetch` and `create_starting_node_tree` are what kind of nodes to use creating the starting node tree. This can be better named certainly - and this will also be changing.

The new API design will be that the `Molecule` is created, but isn't necessarily linked with a 3D object. Calling `mol.create_object()` will create a new 3D object linked to the data.

Like I have said previously, I would love feedback and suggestions on design, and how people would like to be using this. 

The priority for Molecular Nodes is to be a GUI application, aimed to be used via Blender doing pointing and clicking (mostly because Blender is mostly designed this way) - but I would like to better support usage in scripting workflows.

**Comment by nilshof01** [comment]
Hi, 

Thank you for the quick reply! First of all, you do not have to apologize for anything, since this package is insane and I want to thank you for that :)
I get the point with filtering of water atoms and thank you for this detail ! :)
 So, to sum up I guess I should just wait until next week and use your code snipped to update the colors right? - If I do so, will the object then also automatically be updated? My final goal is to just export the object as a fbx file and I wonder if the molecule would have the applied coloring scheme then. 
Regarding your request for suggestions, I am not an expert in neither blender or molecularnodes but it would be nice to just build a class which gives easy control over the attributes and then updates the object. For example, the class would have methods such as change_color which changes the colors in a residue range given an array and then have more advanced features such as change_viz_in_range which changes the visualization type, for instance from sticks to ribbon in  a residue range  x to y. Probably there are more options which may be integrated but this is what came to my mind in that moment.

Thanks a lot!

**Reply by BradyAJohnston** [reply]
I think having more convenience class functions for things like setting the color like you suggest is a good idea. I'll hopefully be able to put together some more documentation once everything is finalised for 4.2
